k5test/realm.py

```diff
@@ -371,75 +420,335 @@ def stop(self):
+                'dns_lookup_kdc': 'false',
+                'dns_lookup_realm': 'false'},
+            'realms': {
+                '$realm': {
+                    'kdc': '$hostname:$port0',
+                    'admin_server': '$hostname:$port1',
+                    'kpasswd_server': '$hostname:$port2'}},
+            'logging': {
+                'kadmind': 'FILE:$tmpdir/kadmind.log',
+                'kdc': 'FILE:$tmpdir/kdc.log',
+                'kpasswdd': 'FILE:$tmpdir/kpasswdd.log',
+                'krb5': 'FILE:$tmpdir/krb5.log',
+                'default': 'FILE:$tmpdir/others.log'},
+            'kdc': {
+                'database': {
+                    'dbname': '$tmpdir/db',
+                    'mkey_file': '$tmpdir/stash',
+                    'acl_file': '$tmpdir/acl',
+                    'log_file': '$tmpdir/db.log'},
+                'ports': '$port0'}}
+
+    @property
+    def _kdc_conf(self):
+        return
+
+    def create_kdb(self):
+        self.run_kadminl(['stash', '--key-file=%s/stash' % self.tmpdir,
+                          '--random-password'])
+        self.run_kadminl(['init', self.realm], input="\n\n")
+
+    def addprinc(self, princname, password=None):
+        args = ['add', '--use-defaults']
+        if password:
+            args.append('--password=%s' % password)
         else:
-            kdc_conf_path = None
-        return self._make_env(krb5_conf_path, kdc_conf_path)
+            args.append('--random-key')
+
+        args.append(princname)
+
+        self.run_kadminl(args)
+
+    def change_password(self, principal, password=None, keysalt=None):
+        args = ['change_password']
+
+        if password:
+            args.append('--password=%s' % password)
+        else:
+            args.append('--random-key')
+
+        if keysalt:
+            args.extend('-e', keysalt)
+
+        args.append(principal)
+        self.run_kadminl(args)
+
+    def extract_keytab(self, princname, keytab):
+        self.run_kadminl(['ext', '--keytab=%s' % keytab, princname])
+
+    def kinit(self, princname, password=None, flags=None, verbose=True,
+              **keywords):
+
+        cmd = [self._kinit]
+
+        input = None
+        if password:
+            input = password + "\n"
+            cmd.append('--password-file=STDIN')
+
+        if flags:
+            cmd.extend(flags)
+        cmd.append(princname)
+
+        return self.run(cmd, input=input, **keywords)
+
+    def klist(self, ccache=None, **keywords):
+        return self.run([self._klist, '-c', ccache or self.ccache],
+                        **keywords)
+
+    def klist_keytab(self, keytab=None, **keywords):
+        return self.run([self._ktutil, '-k', keytab or self.keytab, 'list'],
+                        **keywords)
+
+    def prep_kadmin(self, princname=None, pw=None, flags=None):
+        raise NotImplementedError()  # FIXME
+
+    def run_kadmin(self, query, **keywords):
+        raise NotImplementedError()  # FIXME
```

Comment:
Can you tell me more about what's going on here?

Reply:
I had no need (unused in python-gssapi or my test library) for both of these functions but knew they would be done differently across Heimdal and MIT so made them an abstract method. If someone does need them for some reason they are free to try and implement them.